byoqm/source_repository/source_repository.py

```
@@ -14,55 +24,103 @@ class SourceRepository:
-            )
-            raise e
+        with of_file.file_path.open("rb") as file:
+            ast = self.tree_sitter_parsers[of_file.language].parse(file.read())
+            if ast is None:
+                raise TypeError(f"Abstract syntax tree for ${of_file} is None")
+            return ast
+
+    def _discover_files(self) -> Dict[Path, FileInfo]:
+        if self.src_root.is_file():
+            return {self.src_root: self._inspect_file(self.src_root)}
+
+        file_infos: Dict[Path, FileInfo] = self._discover_in_dir(self.src_root)
+        return file_infos
+
+    def _discover_in_dir(self, root_dir: Path) -> Dict[Path, FileInfo]:
+        file_infos: Dict[Path, FileInfo] = {}
+        for f in root_dir.iterdir():
+            if f.is_dir():
+                file_infos.update(self._discover_in_dir(f))
+            else:
+                file_info = self._inspect_file(f)
+                if not self._should_exclude(file_info):
+                    file_infos[f] = file_info
+
+        return file_infos
+
+    def _inspect_file(self, file_path: Path) -> FileInfo:
+        if not file_path.is_file():
+            raise ValueError(f"_inspect_file expects that ${file_path} is a file")
+
+        programming_language: str = "unknown"
+        if file_path.suffix == ".py":
+            programming_language = "python"
+        elif file_path.suffix == ".cs":
+            programming_language = "c_sharp"
+        elif file_path.suffix == ".java":
+            programming_language = "java"
+
+        encoding: str = "unknown"
+        if file_path.suffix != ".jar":
+            with file_path.open("rb") as file:
+                chardet_guess = chardet.detect(file.read())
+                if not chardet_guess["encoding"] is None:
+                    # encoding will be 'unknown' if chardet guesses it as None
+                    encoding = chardet_guess["encoding"]
```

same as before. If we default strings to an arbitrary value, I would argue that we need some form of error handling

---

Again, unknown isn't an immediate error. But something that indicates that file should be excluded.